# Foundations of Finance

## Homework 1 Solutions

These Notebooks embed text and code in the same document. All cells can be evaluated by pressing shift+enter.

We will start by importing some packages that will aid us with our problems. We will be importing numpy along with some packages from math and scipy library.

In [1]:
# Comments within blocks of code start with a # sign and continue until the end of the line
import numpy as np
from math import log, sqrt, exp
from scipy import stats
from IPython.display import Image
from IPython.core.display import HTML 
import pandas_datareader as pd_dr
import pandas as pd

### Topic 1: Financial Markets

#### 1. You are among the OTC marketmakers in the stock of BioEngineering, Inc. and quote a bid of 102.25 and an ask of 102.50. Suppose that you have a zero inventory.

##### (a) On Day 1 you receive market buy orders for 10,000 shares and market sell orders for 4,000 shares. How much do you earn on the 4,000 shares that you bought and sold? What is the value of your inventory at the end of the day? 
(Hints: It is possible
to have negative inventory. Further, there is more than one correct way to value an
inventory, but please state what assumption your valuation is based on.)

You have sold 10,000 shares at the ask price of 102.50. You bought 4,000 shares at
a bid price 102.25. Thus, 6,000 shares are sold short (sold without already owning
the security). Your revenue from the 4,000 "round trip" purchase and sale produces a
profit equal to the bid minus the ask times the volume done. Hence, the profit on the
round trip trades is 0.25 x 4,000 = 1,000.
The value of your inventory is equal to the value of your short position of 6,000 shares.
Since there is both a bid and an ask price, this question can answered is various ways
depending on what you assume:
The "conservative" valuation is to value your position at the ask price of 102.50. Then,
you have a position of -615,000. This conservative valuation is useful because, if you
cover your short position by buying from another dealer at his ask price of 102.50,
you would have to pay 615,000. (Also, in this example it is the price for which you
sold the securities.)
The "aggressive" valuation is to value your position at the bid price of 102.25. Then,
you have a position of -613,500 (i.e. less negative than above). This value is implicitly
based on an expectation that some investors will come to you and sell you 6000 shares
at your current bid price.
Often, real-world market makers will value their inventory at the mid price, in this
case 102.375. Then, you have a position of -614,250.


##### (b) Before trading begins on Day 2 the company announces trial testing of a cure for acne in mice. The quoted bid and ask jump to 110.25 and 110.50, respectively.

During Day 2 you receive market sell orders for 8,000 shares and buy orders for 2,000
shares. What is your total profit or loss over the two-day period? What is the value
of your inventory at the end of Day 2?
You have bought 8,000 shares during Day 2 at 110.25 and sold 2,000 shares at 110.50.
On the 2,000 you bought and sold during the day you earn 2,000 x 0.25 = 500. You
also added 6,000 shares to your inventory at a price of 110.25. Since you were short
6,000 shares at 102.50 from yesterday's trading, your loss on these 6,000 shares is -7.75 x 6,000 = -46; 500. Thus your total profit/loss over the two-day period is:

                               1,000+500-46; 500 = -45; 000.

Your inventory at the end of Day 2 is zero since
you purchased 6,000 shares that offset the 6,000 share short position at the end of Day
1.

##### (c) What is a market maker's objective? Is there anything you could have done during Day 1, consistent with a market maker's objective, that would have improved your performance over the two-day period?

A marketmaker's objective is to earn the bid-ask spread, and not (necessarily) to
speculate on equilibrium price movements. The 6,000 share short position at the end
of Day 1 left you vulnerable to a jump in quoted prices. Perhaps you should have
increased the prices during Day 1 as you observed more buying than selling. Also,
you could have reduced your short inventory position by buying from other dealers at
102.50 as the market closed on Day 1. There are two factors that may have prevented
you from doing this: (1) Your expected profit of .25 on the 6,000 shares if market
prices remained unchanged at 102.25 (bid) and 102.50 (ask). (2) Unwillingness of
other dealers to accommodate your purchase of 6,000 shares at the 102.50 ask price.

### Topic 2: Performance Measures

#### 2. Suppose a 5-year zero-coupon Treasury bond with face value $1000 has a 5% yield (annually compounded).
##### (a) What price does this bond sell for?

Let's assign values to the parameters of Time value of money.

In [23]:
fv = 1000                  #Face Value
r = 0.05                   #Rate of interest
t = 5                      #Time Period

In [24]:
pv = fv/(1+r)**t           #Formula for calculating the Present Value

In [25]:
pv                         #Calling out the value to print the final result.

783.5261664684588

We can solve the above problem with a single line of code using scipy package

In [26]:
#numpy.pv(rate, nper, pmt, fv=0.0, when='end')

# a future value, fv
# an interest rate compounded once per period, of which there are
# nper total
# a (fixed) payment, pmt, paid either
# at the beginning (when = {‘begin’, 1}) or the end (when = {‘end’, 0}) of each period


np.pv(0.05,5,0,1000)

-783.52616646845877

In financial terms, a negative sign before a cash flow indicates an outgoing cash flow and a positive or no sign indicates an incoming cash flow.



##### (b) Suppose another zero-coupon Treasury bond also has a 5% yield, but sells for $325.57. What is the maturity of this bond?

In [27]:
fv = 1000                  #Face Value
r = 0.05                   #Rate of interest
pv = 325.57                #Present Value

In [28]:
t = log(fv/pv)/log(1+r)

In [29]:
t

23.000082204967935

Using scipy:

In [30]:
#numpy.nper(rate, pmt, pv, fv=0, when='end')

np.nper(0.05,0,-325.57,1000)

C:\Users\tmahendru\Anaconda3\lib\site-packages\numpy\lib\financial.py:281: RuntimeWarning: divide by zero encountered in double_scalars
  A = -(fv + pv)/(pmt+0.0)


23.000082204967935

#### 3. Which of the following investments do you prefer?

##### (a) Purchase a zero-coupon bond, which pays 1000 in ten years, for a price of $550.

##### (b) Invest $550 for ten years in Chase at a guaranteed annual interest rate of 5.5%.

Solution:
You can go about this in two ways :

1. We can calculate the return from the bond and compare that with the interest guaranteed from Chase. Whichever gives us higher returns will be preferred.
Or

2. We can calculate the future value of our investment in Chase and compare that with the face value of our bond.

In [31]:
fv = 1000
t = 10
pv = 550

In [32]:
r = (fv/pv)**(1/t)-1

In [33]:
r

0.06160689621814597

The annual return on the bond is 0.0616 or 6.16% which is greater than 5.5% yielded by investing in Chase. Therefore, the bond is a better investment.

In [34]:
r = 0.055
pv = 550
t = 10

In [35]:
fv = pv*(1+r)**t                 #The variable after '**' signifies the exponent

In [36]:
fv

939.4794520944758

the Chase deposit after 10 years will have grown to $939.48 which is lower than the face value of the bond

Using Scipy:

In [37]:
#numpy.rate(nper, pmt, pv, fv, when='end', guess=0.1, tol=1e-06, maxiter=100)[source]

np.rate(10,0,-550,1000)

0.061606896218165071

In [38]:
np.fv(0.055,10,0,-550)

939.47945209447585

** 4. Suppose you get for free one of following two securities: **

##### (a) an annuity that pays 10000 at the end of each of the next 6 years; or 
##### (b) a perpetuity that pays 10,000 forever, but it does not begin until 10 years from now (the first cash payment from this security is 11 years from today). 

#### Which security would you choose if the annual interest rate is 5%? Does your answer change if the intere3st rate is 10%? Explain why or why not.

To determine whether it is better to get for free (a) or (b), we must calculate which
security has the higher present value.

It is always helpful to show a time line of the cash 
ows (to simplify the picture we
omit the 000s in 10,000).

In [39]:
path = "https://raw.githubusercontent.com/tanmmay92/Finance-TF-Work/master/Timeline.PNG"
Image(url = path, width=500, height=500)

Cash flow (a) is a straightforward annuity whose present value can be calculated as follows: 

In [40]:
C = 10000
r=0.05
t = 6

In [41]:
pv = C*((1/r)-1/(r*(1+r)**t))

In [42]:
pv

50756.92067267454

Cash Flow (b) is a perpetuity that begins 10 years from now. We can value it in two
parts. First, we know that a perpetuity has a value, P, given by: P = C/r

In [43]:
C=10000
r=0.05

In [44]:
P=C/r

In [45]:
P

200000.0

Thus, with C = 10000 and r = 0.05 the perpetuity is worth 200,000. But it is worth
200,000 ten years from now, not today. We can get the present value of 200,000
received with a 10 year delay by treating that sum as though it were a zero coupon
bond with a face value of 200,000 payable in 10 years. Hence,

In [46]:
fv = 200000
r = 0.05
t = 10

In [47]:
pv = fv/(1+r)**t

In [48]:
pv

122782.65070815184

We find that the present value of the
perpetuity beginning 10 years from now is 122,782. Hence, we prefer the perpetuity
because its present value of 122,782 is larger than the 50,757 present value of the
annuity in (a).

Similarly, when rate = 10%, calculating for both:

In [49]:
C = 10000
r=0.1
t = 6

In [50]:
pv = C*((1/r)-1/(r*(1+r)**t))            #For Annuity

In [51]:
pv                                       #For Annuity

43552.60699462229

In [52]:
pv = C/r                                 #Present Value For Perpetuity 10 years from now

In [53]:
pv                                       #Present Value For Perpetuity 10 years from now

100000.0

Similarly, We can get the present value of 100,000 received with a 10 year delay by treating that sum as though it were a zero coupon
bond with a face value of $100,000 payable in 10 years.

In [54]:
fv = 100000
r = 0.1
t = 10

In [55]:
pv = fv/(1+r)**t                         #Present Value For Perpetuity at Day 0

In [56]:
pv                                       #Present Value For Perpetuity at Day 0

38554.32894295314

Thus at 10% we prefer the annuity in (a), with a present value of 43,552, to the
perpetuity, which is worth only $38,554.
This result stems from the fact that even though the perpetuity has infinite cash flows
compared with the annuity, those cash flows begin with a 10- year delay. At a 10%
interest rate the delayed cash 
flows are penalized very heavily.






** 5. Suppose a hedge fund manager earns 1% per trading day. There are 250 trading days
per year. Answer the following questions: **


** (a) What will be your annual return on $100 invested in her fund if she allows you to
reinvest in her fund the 1% you earn each day? **

Allowing you to reinvest at 1% per day means that you are earning compound interest
on your initial $100 investment. The formula for P growing to F for one year at a
compound rate r per annum is:
                             F = P(1+r/n)^n

where n is the number of compounding periods per year and hence r/n is the rate per
compounding period. We are given r/n = 1% per day and are asked to calculate the
annual yield. This is equivalent to asking for the effective annual rate.


In [57]:
r=.01
t = 250
pv=100

In [58]:
F = pv*(1+r)**t

In [59]:
F

1203.2155768297464

#### (b) What will be your annual return assuming she puts all of your daily earnings into a zero-interest- bearing checking account and pays you everything earned at the end of the year?

If the hedge fund manager insists on putting your daily 1% earnings into a zero-interest
bearing checking account, then you will earn only the daily rate (1%) multiplied by
the number of days, or, 1% * 250 = 250%

Notice that this is equivalent to the annual percentage rate (APR) calculation:
APR = periodic rate x n = 1% x 250

The value at the end of the year includes interest earnings plus original investment,
that is, 100+250=350.

#### (c) Can you summarize when it is proper to "annualize" using APR (annual percentage rate) versus EAR (eective annual rate)?

Whether you use APR or EAR to annualize a periodic rate depends upon the process
for reinvesting the proceeds of your investment. If you can reinvest at the periodic
rate (as in (a)) then EAR is appropriate. If the reinvestment rate is zero (as in (b))
then APR is appropriate. Since the reinvestment rate is rarely zero, the APR usually
understates the annual rate.

#### 6. Here are some alternative investments you are considering for one year. (i) Bank A promises to pay 8% on your deposit compounded annually. (ii) Bank B promises to pay 8% on your deposit compounded daily. Compare the effective annual rate (EAR) on these investments.
(i) EAR = .08


In [66]:
r=.08
n=365      #compounded daily


In [67]:
EAR = ((1+r/n)**n)-1

In [68]:
EAR

0.08327757179283468

In general, the greater the compounding frequency, the higher the EAR. This is because
with more frequent compounding, you earn more interest on interest.

#### 7. (a) Suppose that you have purchased a 3-year zero-coupon bond with face value of 1000 and a price of 850. If you hold the bond to maturity, what is your annual rate of return?

In [60]:
pv = 850
fv = 1000
t = 3

In [61]:
r = (fv/pv)**(1/t)-1

In [62]:
r

0.05566719197800074

In [63]:
np.rate(3,0,-850,1000)        #Using Scipy

0.055667191978000838

#### (b) Now suppose you have purchased a 3-year bond with face value of 1000, a 7% annual coupon, and a price of 975. Assuming that you hold the bond to maturity, is the IRR greater or less than the return on the bond in part (a)?

Recall that the IRR is defined as the interest rate that makes the present value of the
payments equal to the price. Hence, the IRR (annual compounding) solves : 

                           975 = 70/(1 + IRR) + 70/(1 + IRR)^2 + 1070/(1 + IRR)^3 

In [64]:
round(np.irr([-975, 70, 70, 1070, 20]), 3)        #Using Scipy

0.085

Another way to go about it is to insert the IRR value calculated in (a), to find out pv, and compare those values.

#### 8. Download the monthly S&P 500 prices from January 1950 until today http://finance.yahoo.com/ (click Market Data, click S&P 500, click Historical Prices, click Monthly, click Get Prices, click Download to Spreadsheet at the bottom of the page).

#### (a) What is your best estimate for next month’s return?

In [2]:
path = 'C:\\Users\\Aditya\\Downloads\\' #Add path to file
file = '^GSPC.csv' #Add file name
pd.df = pd.read_csv(path + file) #Load csv in data frame
dff = pd.df.set_index(['Date']) # Set index on date

In [3]:
dff # Data frame with index set on date

Open         High          Low        Close    Adj Close  \
Date                                                                         
1/1/1950     16.660000    17.090000    16.650000    17.049999    17.049999   
2/1/1950     17.049999    17.350000    16.990000    17.219999    17.219999   
3/1/1950     17.240000    17.610001    17.070000    17.290001    17.290001   
4/1/1950     17.340000    18.070000    17.340000    18.070000    18.070000   
5/1/1950     18.219999    18.780001    18.110001    18.780001    18.780001   
6/1/1950     18.770000    19.400000    17.440001    17.690001    17.690001   
7/1/1950     17.639999    17.910000    16.680000    17.840000    17.840000   
8/1/1950     18.020000    18.820000    17.950001    18.420000    18.420000   
9/1/1950     18.549999    19.450001    18.540001    19.450001    19.450001   
10/1/1950    19.690001    20.160000    19.530001    19.530001    19.530001   
11/1/1950    19.559999    20.320000    19.360001    19.510000    19.510000   
12/1/1950    19.660000    20.430000    19.000000    20.410000    20.410000   
1/1/1951     20.770000    21.740000    20.690001    21.660000    21.660000   
2/1/1951     21.770000    22.209999    21.760000    21.799999    21.799999   
3/1/1951     21.850000    21.950001    21.250000    21.400000    21.400000   
4/1/1951     21.320000    22.430000    21.260000    22.430000    22.430000   
5/1/1951     22.530001    22.809999    21.030001    21.520000    21.520000   
6/1/1951     21.480000    22.049999    20.959999    20.959999    20.959999   
7/1/1951     21.100000    22.629999    21.100000    22.400000    22.400000   
8/1/1951     22.510000    23.280001    22.510000    23.280001    23.280001   
9/1/1951     23.280001    23.709999    23.260000    23.260000    23.260000   
10/1/1951    23.469999    23.850000    22.450001    22.940001    22.940001   
11/1/1951    23.100000    23.100000    22.299999    22.879999    22.879999   
12/1/1951    22.940001    23.770000    22.940001    23.770000    23.770000   
1/1/1952     23.799999    24.660000    23.740000    24.139999    24.139999   
2/1/1952     24.299999    24.410000    23.090000    23.260000    23.260000   
3/1/1952     23.280001    24.370001    23.280001    24.370001    24.370001   
4/1/1952     24.180000    24.180000    23.320000    23.320000    23.320000   
5/1/1952     23.170000    23.940001    23.170000    23.860001    23.860001   
6/1/1952     23.799999    24.959999    23.780001    24.959999    24.959999   
...                ...          ...          ...          ...          ...   
4/1/2015   2067.629883  2125.919922  2048.379883  2085.510010  2085.510010   
5/1/2015   2087.379883  2134.719971  2067.929932  2107.389893  2107.389893   
6/1/2015   2108.639893  2129.870117  2056.320068  2063.110107  2063.110107   
7/1/2015   2067.000000  2132.820068  2044.020020  2103.840088  2103.840088   
8/1/2015   2104.489990  2112.659912  1867.010010  1972.180054  1972.180054   
9/1/2015   1970.089966  2020.859985  1871.910034  1920.030029  1920.030029   
10/1/2015  1919.650024  2094.320068  1893.699951  2079.360107  2079.360107   
11/1/2015  2080.760010  2116.479980  2019.390015  2080.409912  2080.409912   
12/1/2015  2082.929932  2104.270020  1993.260010  2043.939941  2043.939941   
1/1/2016   2038.199951  2038.199951  1812.290039  1940.239990  1940.239990   
2/1/2016   1936.939941  1962.959961  1810.099976  1932.229980  1932.229980   
3/1/2016   1937.089966  2072.209961  1937.089966  2059.739990  2059.739990   
4/1/2016   2056.620117  2111.050049  2033.800049  2065.300049  2065.300049   
5/1/2016   2067.169922  2103.479980  2025.910034  2096.949951  2096.949951   
6/1/2016   2093.939941  2120.550049  1991.680054  2098.860107  2098.860107   
7/1/2016   2099.340088  2177.090088  2074.020020  2173.600098  2173.600098   
8/1/2016   2173.149902  2193.810059  2147.580078  2170.949951  2170.949951   
9/1/2016   2171.330078  2187.870117  2119.120117  2168.270020  2168.270020   
10/1/2016  2164.330078  2169.600098  211

In [16]:
R = [] # Declaring empty list

R = [(dff.iloc[i, 3] - dff.iloc[i-1,3])/dff.iloc[i-1, 3] for i in range(1,812)] #Computing Return for each month

np.mean(R) # Mean of monthly returns

0.0070080464590915395

#### (b) What would have been your annualized HPR if you invested as of the start of the index?

In [17]:
R2 = []
R2 = [(R[i]+1) for i in range(0,811)] # Calculating 1+R

# Function for GeoMean
def geo_mean_overflow(iterable):
    a = np.log(iterable)
    return np.exp(a.sum()/len(a))

# Calculating Annual HPR
np.power(geo_mean_overflow(R2),12)

1.0764137199026858

#### Alternatively,

In [18]:
np.power(np.prod(R2),12/813) #Annual HPR

1.076218752072424

#### (c)  In what month occurred the lowest monthly return? What happened?

In [20]:
idx=R.index(min(R)) #Getting index of lowest return
idx = idx+1

dff.index[idx] #Looking for date corresponding to the index

'10/1/1987'

In [22]:
min(R) #Min Return

-0.21763041614888423

October 1987 (not September 1987!) with a return of -21:763%. This was mainly concentrated on Monday October 19, 1987 known as Black Monday. See e.g.: [Black Monday](http://en.wikipedia.org/wiki/BlackMonday_(1987))